### ISSUE TITLE
[BUG] QLinearConv uses same value for all zero_points!

### ISSUE THREAD

**Issue by mikeseven** [microsoft/onnxruntime#8890]
https://github.com/microsoft/onnxruntime/blob/36fa0de8b729624d09094b22302781cd0e3ebc78/onnxruntime/core/providers/cpu/nn/qlinearconv.cc#L247

Zero-points as 1D tensor have different values for each element when quantized per output channel.
When W is quantized asymmetric, this is of course the case.

ORT quantization tool can generate ONNX files with asymmetric weights but runtime crashes here and shows this error.

Thanks for fixing this bug, I really need this fix to run models that otherwise have very bad accuracy.

**Comment by chenfucn** [comment]
This is not a bug but  a feature limitation as per channel quantization is currently not supported in quantized linear convolution, with checking logic and error messages indicating as such.

**Comment by yufenglee** [comment]
@mikeseven, per-channel quantization for QLinearConv is supported when W is quantized symmetrically. Like chen said, it is not bug. For the bad accuracy, which cpu are you running on? Did you try turning on reduce_range option https://onnxruntime.ai/docs/how-to/quantization.html#when-to-use-per-channel-and-reduce-range?

**Comment by mikeseven** [comment]
Thanks for your reply.
This limitation is not mentioned anywhere in the documentation. I would still call it a bug because the quantization tool generates models without this limitation, which don't run on the engine, with error "zero point must be constant". This happens with weight symmetric = false extra options in quantize_static, for example.

The precision I need is not related to overflow. It is related to channel distributions rather skewed where a symmetric range creates huge errors.

**Comment by chenfucn** [comment]
Thanks for pointing out that our documentation needs improvement. 

W.r.t to quantization tool does not catch the problem, on the other hand...

Unfortunately, it is common for a software program to evaluate a partial function, where locating the exact domain for said function is more complex than evaluating the function itself. This makes it unrealistic to construct a separate input checker. Our quantization tool does have some sanity check but it is by no means a comprehensive input validator.

**Comment by mikeseven** [comment]
what the quantization tool should do for this limitation is force all zp to be the same in the 1D tensor case. Not ideal but at least consistent. 
Alternatively, the feature could be disabled until it's ready. It's better to fail early than, in my case, spend hours figuring out what was wrong....

**Comment by rlewkowicz** [comment]
This is a pain for me too. @chenfucn @yufenglee 

Im going to sound ignorant. I don't understand the underlying dynamics like you do, but as an outcome, it's not functional in the way I need it to be. 

I too am struggling with accuracy. So what I do is I go pull all the observers from my QAT pt weights and heuristically match them to the desired nodes. For activation, this is fine. The pytorch observers are per channel, asymmetric.  For weights this is a problem. 

And it's not just a small accuracy dip. I can't believe people don't talk about this. Min/Max is worthless. Every blog, every tool, everything relies on min/max and it's garbage. It doesn't work. It's not just a small dip, it's a brutal one. Entropy or percentile cause OOM VERY FAST. I have 120G of memory and it ooms almost instantly.  I had gone down the road of building my own entropy calc and data loaders and all that. It's such a pain. Then newer onnx versions don't seem to expose things as well as the older versions so I bailed on that. 

I like my PT->onnx heuristics match. But that doesn't help me with the weights.

**Comment by rlewkowicz** [comment]
It turns out it doesn't matter. I think it's because onnx only supports per layer? At least on the output logits. Because even if you do rewrite it, it can only take a single channel zp. 

```
2025-08-13 14:32:55.951558885 [E:onnxruntime:, inference_session.cc:2105 operator()] Exception during initialization: /onnxruntime_src/onnxruntime/contrib_ops/cpu/quantization/qlinear_lookup_table.cc:52 void onnxruntime::contrib::QlinearBuildLookupTable(uint8_t*, const onnxruntime::Tensor*, const onnxruntime::Tensor*, const onnxruntime::Tensor*, const onnxruntime::Tensor*, const LookupTableArrayTransformer&) [with T = unsigned char; uint8_t = unsigned char; LookupTableArrayTransformer = std::function<void(const float*, float*, long unsigned int)>] IsScalarOr1ElementVector(tensor_y_scale) was false. QlinearBuildLookupTable : input Y_scale must be a scalar or 1D tensor of size 1
```
